okpt/test/profile.py

```diff
@@ -0,0 +1,146 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Provides decorators to profile functions.
+
+The decorators work by adding a `measureable` (time, memory, etc) field to a
+dictionary returned by the wrapped function. So the wrapped functions must
+return a dictionary in order to be profiled.
+"""
+import functools
+import time
+from typing import Callable, Dict
+
+import psutil
+
+
+def step(f: Callable):
+    """Decorates a function as a test step.
+
+    Because the functions to profile a step require a dictionary as the step
+    output, this function, besides semantically denoting a function as a step,
+    also wraps the step and forces the output to be an empty dictionary if
+    needed.
+
+    Args:
+        f: Function to decorate.
+
+    Returns:
+        If the output of the passed in function is a dictionary, it returns that
+        output. Otherwise, it returns an empty dictionary.
+    """
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        result = f(*args, **kwargs)
+        if isinstance(result, dict):
+            return result
+        else:
+            return {}
+
+    return wrapper
+
+
+class _Timer():
+    """Timer class for timing.
+
+    Methods:
+        start: Starts the timer.
+        end: Stops the timer and returns the time elapsed since start.
+    """
+    def __init__(self):
+        self.start_time = time.perf_counter()
+
+    def start(self):
+        """Starts the timer."""
+        self.start_time = time.perf_counter()
+
+    def end(self) -> float:
+        """Stops the timer.
+
+        Returns:
+            The time elapsed in milliseconds.
+        """
+        return (time.perf_counter() - self.start_time) * 1000
+
+
+def memory(f: Callable[..., Dict]):
+    """Profiles a functions memory usage.
```

Comment:
Memory usage is kind of vague. Could you add more detail on what memory you are actually profiling?

Also, this will not work on remote processes. 

Also, this checks the whole systems memory usage. Im am wondering what the variability in this measure is inside a container and outside the container?

Reply:
Since profiling memory introduces a lot of different complexities, I've removed it from this PR.